GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.DownloadObject.cs

```diff
@@ -0,0 +1,165 @@
+        // See issue #21.
+
+        public void DownloadObject(string bucket, string sourceName, Stream destination)
+        {
+            DownloadObject(bucket, sourceName, destination, options: null, progress: null);
+        }
+
+        public void DownloadObject(
+            string bucket,
+            string sourceName,
+            Stream destination,
+            DownloadObjectOptions options,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(bucket, sourceName);
+            DownloadObjectImpl(uri, destination, options, progress);
+        }
+
+        public Task DownloadObjectAsync(string bucket, string sourceName, Stream destination)
+        {
+            return DownloadObjectAsync(bucket, sourceName, destination,
+                options: null, cancellationToken: CancellationToken.None, progress: null);
+        }
+
+        public Task DownloadObjectAsync(
+            string bucket,
+            string sourceName,
+            Stream destination,
+            DownloadObjectOptions options,
+            CancellationToken cancellationToken,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(bucket, sourceName);
+            return DownloadObjectAsyncImpl(uri, destination, options, cancellationToken, progress);
+        }
+
+        public Task DownloadObjectAsync(Object source, Stream destination)
+        {
+            return DownloadObjectAsync(source, destination,
+                options: null, cancellationToken: CancellationToken.None, progress: null);
+        }
+
+        public void DownloadObject(Object source, Stream destination)
+        {
+            DownloadObject(source, destination, options: null, progress: null);
+        }
+
+        public void DownloadObject(
+            Object source,
+            Stream destination,
+            DownloadObjectOptions options,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(source);
+            DownloadObjectImpl(uri, destination, options, progress);
+        }
+
+        public Task DownloadObjectAsync(
+            Object source,
+            Stream destination,
+            DownloadObjectOptions options,
+            CancellationToken cancellationToken,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(source);
+            return DownloadObjectAsyncImpl(uri, destination, options, cancellationToken, progress);
+        }
+
+        /// <summary>
+        /// Constructs the media URL of an object from its bucket and name.
+        /// </summary>
+        private static string GetUri(string bucket, string sourceName)
+        {
+            bucket.CheckNotNull(nameof(bucket));
+            sourceName.CheckNotNull(nameof(sourceName));
+            return $"https://www.googleapis.com/download/storage/v1/b/{Uri.EscapeDataString(bucket)}/o/{Uri.EscapeDataString(sourceName)}?alt=media";
```

Comment:
The way I've implemented it (in a later revision), we do very basic validity checks - enough to validate that we don't need to escape it - but that's all. It's meant to be sufficiently lenient that there's no way the server would accept a name that we don't, but strict enough to make sure we don't need to escape it.  See https://github.com/GoogleCloudPlatform/gcloud-dotnet/blob/master/GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.cs#L73 for the regex it uses. Basically it validates the first bullet of https://cloud.google.com/storage/docs/bucket-naming?hl=en#requirements and part of the third:

> Bucket names must contain only lowercase letters, numbers, dashes (-), underscores (_), and dots (.).
> ...
> ...  Names containing dots can contain up to 222 characters ...

That's all we validate, because it's all we need to validate to avoid escaping. I haven't implemented any of the more stringent length checks, or starting and ending with a number or letter, etc - I just wanted it to be simple enough to provide the guarantee I want around escaping.

Happy to discuss further, but I'm currently content with the balance - it's not niggling me.